File: Sliding_Window/3_longest_substring_without_repeating_characters/3.py



class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        window = [0] * 96
        index = lambda x: ord(x) - 33
        max_count = 0
        count = 0
        i = 0
        j = 0
        while j < len(s):
            if i == j or window[index(s[j])] == 0:
                window[index(s[j])] = 1
                count += 1
                max_count = max(count, max_count)
                j += 1
            else:
                window[index(s[i])] = 0
                count -= 1
                i += 1
        
        return max_count

File: Sliding_Window/3_longest_substring_without_repeating_characters/3.rs


use std::cmp::max;

impl Solution {
    
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut window: Vec<bool> = vec!(false;96);
        let s: Vec<char> = s.chars().collect();
        let mut max_count: usize = 0;
        let mut count: usize = 0;

        fn index(chr: &char) -> usize {
            return (*chr as u32 - 32) as usize;
        }

        let mut i = 0;
        let mut j = 0;
        while j < s.len() {
            
            if (i == j) || (window[index(&s[j])] == false) {
                window[index(&s[j])] = true;
                count += 1;
                max_count = max(count, max_count);
                j += 1;
            } else {
                window[index(&s[i])] = false;
                count -= 1;
                i += 1;
            }
        }
        return max_count as i32;
    }
}